tests/initcoder_test.go

```diff
@@ -0,0 +1,170 @@
+package tests
+
+import (
+	"context"
+	"math/big"
+	"testing"
+
+	"github.com/Fantom-foundation/go-opera/tests/contracts/initcode"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/stretchr/testify/require"
+)
+
+// this constant comes from  https://eips.ethereum.org/EIPS/eip-3860#parameters
+const MAX_INIT_CODE_SIZE uint64 = 49152
+const sufficientGas = uint64(100_000)
+
+func TestInitCodeSizeLimitAndMetered(t *testing.T) {
+	require := require.New(t)
+
+	net, err := StartIntegrationTestNet(t.TempDir())
+	require.NoError(err)
+	defer net.Stop()
+
+	// Deploy the invalid start contract.
+	contract, receipt, err := DeployContract(net, initcode.DeployInitcode)
+	require.NoError(err)
+	require.Equal(types.ReceiptStatusSuccessful, receipt.Status, "failed to deploy contract")
+
+	// deploy measureGasAndAssign to get cost of deploying a contract without create.
+	_, gasUsed := createAndCheckContractWithCodeLenAndGas(t, net, contract, contract.MeasureAssignGasCost, 0, 0)
+
+	// -- using CREATE instruction
+	const wordCostCreate uint64 = 2
+	var gasForCreate uint64 = 32000 + gasUsed
+	t.Run("create", func(t *testing.T) {
+		testForVariant(t, net, contract, contract.CreateContractWith, gasForCreate, wordCostCreate)
+	})
+
+	// -- using CREATE2 instruction
+	// create2 has extra costs for hashing, this is explained in
+	// https://eips.ethereum.org/EIPS/eip-3860 and reflected in evm.codes calculator.
+	const wordCostCreate2 uint64 = wordCostCreate + 6
+	var gasForCreate2 uint64 = gasForCreate + 44
+	t.Run("create2", func(t *testing.T) {
+		testForVariant(t, net, contract, contract.Create2ContractWith, gasForCreate2, wordCostCreate2)
+	})
+
+	t.Run("create transaction", func(t *testing.T) {
+		t.Run("charges depending on the init code size", func(t *testing.T) {
+			// transactions charge 4 gas for each zero byte in data.
+			const zeroByteCost uint64 = 4
+			// create a transaction with 1 word of code.
+			receipt1, err := runTransactionWithCodeSizeAndGas(t, net, 1, sufficientGas)
+			require.NoError(err)
+			require.Equal(types.ReceiptStatusSuccessful, receipt1.Status,
```

Comment:
If I understand the `require` library correctly, you should re-create a new `require` instance in every sub-test because otherwise `require` would report issues to the wrong `t *testing.T`.